clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,148 @@
+        { kernel_param_kind_t::kind_std_layout, 8, 4 },
+        { kernel_param_kind_t::kind_accessor, 4062, 12 },
+
+        { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
+      };
+    } // namespace detail
+  } // namespace _V1
+} // namespace sycl
+```
+
+For each kernel, a mangled name associated with it, is stored in the kernel_names array.
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int `i` and the struct `test_s`), the third one, although a standard layout type, is a type which needs special handling.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+Some of the encoding that we use are:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
+
+# Integration Footer
+An integration footer is a source file produced during device compilation that is intended to be post-included during host compilation of the same source file.
+
+The SYCL 2020 Specification defines a specialization constant as a constant variable where the value is not known until compilation of the SYCL kernel function.
+
+Consider the following example:
+```
+  constexpr specialiation_id<int> int_const;
+  class Wrapper {
+  public:
+    static constexpr specialization_id<float> float_const;
+  };
+```
+
+Also note that the SYCL header files declare, but not define, a special function to obtain the numeric id of a specialization constant.
+```
+     namespace detail {
+         template<auto &SpecConstName>
+         inline const char *get_spec_constant_symbolic_ID();
+     }
+```
+
+
+Partial specializations of this function are defined in the integration footer file:
+```
+   namespace detail {
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<int_const>() {
+       return "unique_name_for_int_const";
+     }
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
+       return "unique_name_for_Wrapper_float_const";
+     }
+   }
+```
+
+This footer file is appended at the end of the translation unit for the host compilation.
+
+Note that unlike for a header file, which has established support for inclusion (e.g., via -include or /FI), there is no clear solution for footer files.  In our implementation, where the footer file is appended to the Translation Unit during host compilation, we have noticed some issues with debugging as well as with ccache.
```

Comment:
Can we turn this into a specific proposal that we can get feedback on? For example:
> Support for a footer file will be provided by passing a temporary file to the host compiler that has `#include` directives for both the primary source file and the footer file (and the header file or do we always want to rely on a `-include`-like feature?). This approach has some drawbacks. For example, some debuggers are sensitive to the name of the primary source file (in what way?) and tools like `ccache` depend on a stable name for the primary source file.

Reply:
Please review.  I have made an attempt at adding details.

Reply:
That looks good to me! Thanks, Prem!